### PR TITLE
Convert rum datalayer

### PR DESCRIPTION
**Pull request by chicharr**
Please always provide the [GitHub issue(s)](../issues) your PR is for, as well as test URLs where your change can be observed (before and after):

Fix #<gh-issue-id>

Test URLs:
- Before: https://convert-rum--bamboohr-website--trieloff.hlx.page/pricing
- After: https://convert-rum-datalayer--bamboohr-website--trieloff.hlx.page/pricing

### CODE REVIEW

File: scripts/scripts.js [hunk 1]
```diff
@@ -14,8 +14,9 @@
  * log RUM if part of the sample.
  * @param {string} checkpoint identifies the checkpoint in funnel
  * @param {Object} data additional data for RUM sample
+ * @param {Object} extraData optional data for analytics tracking
  */
-export function sampleRUM(checkpoint, data = {}) {
+export function sampleRUM(checkpoint, data = {}, extraData = {}) {
```
Comment (trieloff) [inline scripts/scripts.js:19]: I really don't like the signature change in the function. But I'm only at the beginning of the review.

File: scripts/scripts.js [hunk 2]
```diff
@@ -71,7 +72,7 @@ export function sampleRUM(checkpoint, data = {}) {
       sendPing(data);
       if (sampleRUM.cases[checkpoint]) { sampleRUM.cases[checkpoint](); }
     }
-    if (sampleRUM.always[checkpoint]) { sampleRUM.always[checkpoint](data); }
+    if (sampleRUM.always[checkpoint]) { sampleRUM.always[checkpoint](data, extraData); }
```
Comment (trieloff) [inline scripts/scripts.js:75]: I think this could be squeezed into `data` easily. We know what `data` contains, so the laziest thing would be `data.extra` – but I'm sure we can do better.

File: scripts/scripts.js [hunk 3]
```diff
@@ -396,11 +397,13 @@ export function decorateSections($main) {
       const meta = readBlockConfig(sectionMeta);
       const keys = Object.keys(meta);
       keys.forEach((key) => {
-        const styleValues = meta.style.split(',').map((t) => t.trim());
-        styleValues.forEach((style) => {
-          if (key === 'style') section.classList.add(toClassName(style));
-          else section.dataset[key] = meta[key];
-        });
+        if (key === 'style') {
```
Comment (trieloff) [inline scripts/scripts.js:400]: What does this change do? It seems unrelated.

File: scripts/scripts.js [hunk 4]
```diff
@@ -1076,9 +1079,63 @@ async function loadLazy(doc) {
...
+}
+
+/**
+ * Registers conversion listeners according to the metadata configured in the document.
+ * @param {Element} parent element where to find potential event conversion sources
+ * @param {string} path fragment path when the parent element is coming from a fragment
+ */
+export async function registerConversionListeners(parent, path) {
+
+  const conversionElements = getMetadata('conversion-element')?.split(',').map((ce) => toClassName(ce.trim()));
```
Comment (trieloff) [inline scripts/scripts.js]:
```suggestion
  const conversionElements = getMetadata('conversion-element') ? getMetadata('conversion-element').split(',').map((ce) => toClassName(ce.trim())) : [];
```

Optional chaining is not yet supported widely enough for client side code, so we tend to avoid it.

Comment (trieloff) [reply scripts/scripts.js]: You might also want to `toLowerCase` everything as part of the `map`

File: scripts/scripts.js [hunk 5]
```diff
@@ -1076,9 +1079,63 @@ async function loadLazy(doc) {
...
+  if (conversionElements && Array.isArray(conversionElements)) {
```
Comment (trieloff) [inline scripts/scripts.js]:
```suggestion
  if (Array.isArray(conversionElements)) {
```

`Array.isArray(undefined) === false`

File: scripts/scripts.js [hunk 6]
```diff
@@ -1076,9 +1079,63 @@ async function loadLazy(doc) {
...
+    // Track form submissions conversions
+    if (conversionElements.includes('form')) {
```
Comment (trieloff) [inline scripts/scripts.js]: I'd probably turn this list of `if`s into an object like this:

```
const conversionListeners = {
  form: () => { …},
  link: () => {…},
  'labeled-link': () => {…}
}
```

so you can just `filter` the `Object.keys(conversionListeners)`  by `conversionElements.includes(listener)` and then `forEach` run the remaining functions.

File: scripts/scripts.js [hunk 7]
```diff
@@ -1272,18 +1329,35 @@ if (params.get('performance')) {
  * Registers the 'convert' function to `sampleRUM` which sends
  * variant and convert events upon conversion.
  */
-sampleRUM.drain('convert', (elements, cevent, cvalue) => {
+sampleRUM.drain('convert', (elements, cevent, cvalue, extraData) => {
```
Comment (trieloff) [inline scripts/scripts.js]: Let's see what the `extraData` will do 😄

File: scripts/scripts.js [hunk 8]
```diff
@@ -1272,18 +1329,35 @@ if (params.get('performance')) {
...
+        element.addEventListener('click', () => {          
+          const evtDataLayer = {
+            event: "Link Click",
+            eventData: {
+              linkName: cevent,
+              linkText: element.innerHTML,
+              linkHref: element.href              
+            }
+          };
+          sampleRUM.convert(element, cevent, cvalue, evtDataLayer);
```
Comment (trieloff) [inline scripts/scripts.js]: We should revert this bit. The `evtDataLayer` stuff is application-level and customer-dependent. I see the `sampleRUM.convert` as library code that should not have customer specifics.

File: scripts/scripts.js [hunk 9]
```diff
@@ -1299,20 +1373,18 @@ sampleRUM.drain('convert', (elements, cevent, cvalue) => {
...
-sampleRUM.always.on('convert', ({ source, target }) => window.adobeDataLayer && window.adobeDataLayer.push({
-  // TODO: this is just me guessing what the event should be called, same for the payload
-  event: 'Conversion',
-  conversion: {
-    thingClickedOn: source,
-    valueOfAClick: target,
+sampleRUM.always.on('convert', ( data, extraData ) => {
+  console.debug('push to datalayer - convert ', extraData);
+  if(window.digitalData) {
+    window.digitalData.push(extraData);
```
Comment (trieloff) [inline scripts/scripts.js]: Put the actual objects here 😄

**Review by trieloff — CHANGES_REQUESTED** [review]
Good start, let's keep going.

File: scripts/scripts.js (hunk 1 above)
Comment (chicharr) [reply scripts/scripts.js:19]: I implemented it this way because the full `data` object is converted to a json string and sent as the beacon body.
As the extra data is only needed for analytics and not for RUM, I extracted it to a separate object.
The alternative, would be to add to the beacon only the `data` properties that are relevant for RUM, i.e. `source, target`

File: scripts/scripts.js (hunk 2 above)
Comment (chicharr) [reply scripts/scripts.js:75]: see previous comment

File: scripts/scripts.js (hunk 3 above)
Comment (chicharr) [reply scripts/scripts.js:400]: this is fixing a bug introduced by the customer, which prevented section metadata properties to be added as data attributes

File: scripts/scripts.js (hunk 4 above)
Comment (chicharr) [reply scripts/scripts.js]: thanks for the suggestion, I will apply it.
`toClassName` is already doing `toLowerCase`

File: scripts/scripts.js (hunk 9 above)
Comment (chicharr) [reply scripts/scripts.js]: sounds good. I can replace `extraData` by the actual `Element` and generate the datalayer event objects here.

File: scripts/scripts.js (hunk 6 above)
Comment (chicharr) [reply scripts/scripts.js]: Done.

File: scripts/scripts.js (hunk 8 above)
Comment (chicharr) [reply scripts/scripts.js]: Done

File: scripts/scripts.js (hunk 1 above)
Comment (chicharr) [reply scripts/scripts.js:19]: Finally replaced `extraData` by the property `element` in `data`

File: scripts/scripts.js (hunk 5 above)
Comment (chicharr) [reply scripts/scripts.js]: Done
